Update documentation for floating-point precision and determinant calculations

amitsubhashchejara opened the pull request:
This pull request updates the documentation related to floating-point precision in NumPy, specifically addressing the issue of incorrect determinant calculations for certain matrices.

- Added a note on how the determinant of the matrix np.array([[5, 5, 6], [7, 7, 5], [4, 4, 8]]) is expected to be zero, highlighting the observed precision issues.

- Included a code example demonstrating how to compute the determinant and check if it is close to zero using np.isclose().

- Ensured consistency in formatting and improved clarity throughout the documentation.

This change aims to provide users with a better understanding of floating-point arithmetic in NumPy and how to handle potential precision errors.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

amitsubhashchejara committed:
Update basics.types.rst: Clarify determinant precision handling

This update adds clarification to the section on numerical data types regarding the precision of matrix determinants in NumPy. A note has been included explaining that results close to zero, such as those from the determinant of the matrix np.array([[5,5,6],[7,7,5],[4,4,8]]), can appear as very small negative numbers due to floating-point precision issues.

Additionally, a reference to the `np.isclose()` function has been added to guide users on checking if a value is close to zero, which can help mitigate confusion regarding the expected result of zero for certain matrix determinants.

amitsubhashchejara committed:
Update basics.types.rst: Clarify determinant precision handling

This update adds clarification to the section on numerical data types regarding the precision of matrix determinants in NumPy. A note has been included explaining that results close to zero, such as those from the determinant of the matrix np.array([[5,5,6],[7,7,5],[4,4,8]]), can appear as very small negative numbers due to floating-point precision issues.

Additionally, a reference to the `np.isclose()` function has been added to guide users on checking if a value is close to zero, which can help mitigate confusion regarding the expected result of zero for certain matrix determinants.

amitsubhashchejara committed:
Update basics.types.rst: Clarify determinant precision handling

This update adds clarification to the section on numerical data types regarding the precision of matrix determinants in NumPy. A note has been included explaining that results close to zero, such as those from the determinant of the matrix np.array([[5,5,6],[7,7,5],[4,4,8]]), can appear as very small negative numbers due to floating-point precision issues.

Additionally, a reference to the `np.isclose()` function has been added to guide users on checking if a value is close to zero, which can help mitigate confusion regarding the expected result of zero for certain matrix determinants.

amitsubhashchejara committed:
Update basics.types.rst: Clarify determinant precision handling

This update adds clarification to the section on numerical data types regarding the precision of matrix determinants in NumPy. A note has been included explaining that results close to zero, such as those from the determinant of the matrix np.array([[5,5,6],[7,7,5],[4,4,8]]), can appear as very small negative numbers due to floating-point precision issues.

Additionally, a reference to the `np.isclose()` function has been added to guide users on checking if a value is close to zero, which can help mitigate confusion regarding the expected result of zero for certain matrix determinants.

amitsubhashchejara committed:
Update basics.types.rst: Clarify determinant precision handling

This update adds clarification to the section on numerical data types regarding the precision of matrix determinants in NumPy. A note has been included explaining that results close to zero, such as those from the determinant of the matrix np.array([[5,5,6],[7,7,5],[4,4,8]]), can appear as very small negative numbers due to floating-point precision issues.

Additionally, a reference to the `np.isclose()` function has been added to guide users on checking if a value is close to zero, which can help mitigate confusion regarding the expected result of zero for certain matrix determinants.

amitsubhashchejara committed:
Update basics.types.rst: Clarify determinant precision handling

This update adds clarification to the section on numerical data types regarding the precision of matrix determinants in NumPy. A note has been included explaining that results close to zero, such as those from the determinant of the matrix np.array([[5,5,6],[7,7,5],[4,4,8]]), can appear as very small negative numbers due to floating-point precision issues.

Additionally, a reference to the `np.isclose()` function has been added to guide users on checking if a value is close to zero, which can help mitigate confusion regarding the expected result of zero for certain matrix determinants.

amitsubhashchejara committed:
Update basics.types.rst: Clarify determinant precision handling

This update adds clarification to the section on numerical data types regarding the precision of matrix determinants in NumPy. A note has been included explaining that results close to zero, such as those from the determinant of the matrix np.array([[5,5,6],[7,7,5],[4,4,8]]), can appear as very small negative numbers due to floating-point precision issues.

Additionally, a reference to the `np.isclose()` function has been added to guide users on checking if a value is close to zero, which can help mitigate confusion regarding the expected result of zero for certain matrix determinants.

mattip commented:
Closes #27554

A few general comments:
- Please reflow the text to no more than 88 columns
- Use `[skip azp][skip actions][skip cirrus]` in your commit message to limit CI to only documentation runs. You can also use `spin docs` to build the docs locally.
- It would be nice to link to more information for those curious about floating point number representation. For instance, there is [this general reference](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html)
- I am not sure this is the most "discoverable" place to put this blurb. If not moving it to the [top-level linalg documentation](https://numpy.org/doc/stable/reference/routines.linalg.html), what do you think about a reference there "For information about precision in calculations, see ..." ?

amitsubhashchejara committed:
Updating basics.types.rst: Clarify precision handling and link to floating point arithmetic reference. [skip azp][skip actions][skip cirrus]

Updating basics.types.rst:

- Reflowed text to adhere to the 88-column limit for better readability.
- Added a link to an external resource for further information on floating-point arithmetic.

[skip azp][skip actions][skip cirrus]

amitsubhashchejara committed:
Updating basics.types.rst: Clarify precision handling and link to floating point arithmetic reference. [skip azp][skip actions][skip cirrus]

Updating basics.types.rst:

- Reflowed text to adhere to the 88-column limit for better readability.
- Added a link to an external resource for further information on floating-point arithmetic.

[skip azp][skip actions][skip cirrus]

amitsubhashchejara commented:
Thank you @mattip  for the feedback! I’ve made the changes to reflow the text and updated the commit message as suggested.

Regarding the placement of the floating-point precision note, do you think it would be more appropriate to move it to the top-level linalg documentation, because the note I have added is already in basics.types.rst?

Looking forward to your guidance!

rkern commented:
I'm not 100% concerned about discoverability. No one who needs to read it will do so before they try their example and get puzzled, no matter where it is. The main benefit is to have a thing that we, issue responders, can point to after the fact to answer people's questions definitively.

That said, a note in `np.linalg` pointing to the entry is still a good idea; it's not obvious that all of our linear algebra functions are intrinsically floating point.

jakevdp commented on doc/source/user/basics.types.rst:
```diff
@@ -342,6 +342,34 @@ range of possible values.
     >>> np.power(100, 100, dtype=np.float64)
     1e+200
 
+.. note::
```
Rather than putting this in a "note" under "overflow errors", I think it should be main text under a new section called "floating point precision".

amitsubhashchejara replied:
Thank you for the feedback!

I’ve moved the floating-point precision details into a new section titled "Floating point precision" as suggested.

Please let me know if further changes are needed!

amitsubhashchejara committed:
Update basics.types.rst: Added a new section "Floating-point precision". [skip azp][skip actions][skip cirrus]

- Moved the floating-point precision details into a new dedicated section titled "Floating Point Precision" for better discoverability.
- Reflowed text to adhere to the 88-column limit for readability.
- Added a link to an external resource on floating-point arithmetic for users seeking more detailed information.

[skip azp][skip actions][skip cirrus]

jakevdp commented on doc/source/user/basics.types.rst:
```diff
@@ -342,6 +342,35 @@ range of possible values.
+represent decimal numbers. For instance, when computing the determinant of the 
+following matrix:
+
+    >>> import numpy as np
```
You could maybe drive this home even more with a simpler example. My favorite is
```python
>>> 0.3 - 0.2 - 0.1
-2.7755575615628914e-17
```
Though this uses Python float arithmetic rather than NumPy.

jakevdp commented on doc/source/user/basics.types.rst:
```diff
@@ -342,6 +342,35 @@ range of possible values.
+    
+
+One may receive a result like ``-3.1974423109204565e-14`` instead of ``0``. This is
+a known behavior of floating-point operations in numerical libraries.
```
```suggestion
a behavior common to all frameworks that use floating point arithmetic.
```

amitsubhashchejara committed:
Update doc/source/user/basics.types.rst: Update floating-point precision section for clarity.[skip azp][skip actions][skip cirrus]

- Updated the description of floating-point behavior for improved clarity.
- Changed phrasing from "a known behavior of floating-point operations in numerical libraries" to "a behavior common to all frameworks that use floating point arithmetic."

[skip azp][skip actions][skip cirrus]

Co-authored-by: Jake Vanderplas <jakevdp@gmail.com>

amitsubhashchejara commented:
Thank you for the suggestion @jakevdp. I have made the changes as suggested.

eric-wieser commented on doc/source/user/basics.types.rst:
```diff
@@ -342,6 +342,35 @@ range of possible values.
+One may receive a result like ``-3.1974423109204565e-14`` instead of ``0``. This is
+a behavior common to all frameworks that use floating point arithmetic.
+
+To handle such cases, it's advisable to set a threshold for comparison when checking
+if a determinant is effectively zero. For example:
+
+    >>> np.isclose(np.linalg.det(a), 0)  # Check for closeness to 0
+    True
```
This example doesn't actually set a threshold though! I think it might be better to recommend using `np.linalg.cond` instead?

jakevdp replied:
I don't think diving into the details of matrix condition numbers is appropriate for the context here. This is not a discussion of linear algebra concepts, this is a brief section alerting new users that floating point arithmetic is not exact.

jakevdp replied:
My suggestion would be to remove all mentions of the determinant, and show the example I mentioned above, ie.. `0.3 - 0.2 - 0.1 != 0`.

eric-wieser replied:
> I don't think diving into the details of matrix condition numbers is appropriate for the context here.

Good point, I thought I was looking at the docstring for `np.linalg.det`. Perhaps we should link to both here and `cond` from `det` though, if we don't already?

amitsubhashchejara replied:
Yeah, that's what I thought.
So I have updated the example to reflect a more general case of floating-point arithmetic as per your feedback. I removed the determinant-related example and replaced it with the simpler arithmetic case (0.3 - 0.2 - 0.1). Please let me know if further adjustments are needed!

rkern replied:
While that example is important, it demonstrates just one issue with floating point arithmetic (not exactly representing the base-10 decimals used in the literals).

The originating issue that spawned this PR has a very different floating point concern, though, that remains unanswered by the current text. The root cause is that the underlying implementation of `det()` uses a particular floating point algorithm, one which will give an inexact floating point result even if the inputs were exactly representable in floating point arithmetic, which they were in the originating bug report (small integers). If `det()` had been implemented using the high-school pen-and-paper algorithm for determinants, even in floating point arithmetic, it would have given the exact expected answer.

amitsubhashchejara replied:
@rkern would you suggest expanding this section to discuss algorithm-specific floating-point issues, or would it be more appropriate to address that in a different section of the documentation?

rkern replied:
It's not really algorithm-specific concerning just `det()`, which is why I recommended a general note here. It's a general issue for lots of numpy routines. One needs to know that due to several different properties of floating point representations and floating point arithmetic and implementation choices, simple contrived test cases with known exact answers won't always show those exact answers, and this is not a bug.

amitsubhashchejara replied:
Should I add another general note to `np.det()`? Or would you like to suggest a better place for it. Your suggestion would be much appreciated.

rkern replied:
Here.

amitsubhashchejara replied:
So I will open a new PR to add a general note regarding `np.det()` in the "floating point precision" section.

rkern replied:
Not regarding `det()` per se. As I said, this is a general issue. People put in contrived test cases to lots of our functions (especially those in `np.linalg`) that have known exact answers and are surprised when the floating point implementation gives them something inexact. You may use it as an example of the general issue, though.

amitsubhashchejara committed:
Update floating-point precision section: Replaced np.isclose() with np.linalg.cond()

- Replaced the usage of np.isclose() with np.linalg.cond() to assess matrix stability.
- Cond function provides a more accurate measure of how close a matrix is to being singular.

amitsubhashchejara committed:
Update floating-point precision section: Replaced np.isclose() with np.linalg.cond() [skip azp][skip actions][skip cirrus]

- Replaced the usage of np.isclose() with np.linalg.cond() to assess matrix stability.
- Cond function provides a more accurate measure of how close a matrix is to being singular.
[skip azp][skip actions][skip cirrus]

amitsubhashchejara commented:
Thank you for the suggestion @eric-wieser! I've updated the section to use `np.linalg.cond()` instead of `np.isclose()` as you recommended. This should now provide a better indication of matrix stability. Please let me know if any further adjustments are needed.

amitsubhashchejara committed:
Simplified floating-point precision example by removing determinant and adding basic arithmetic case. [skip azp][skip actions][skip cirrus]

- Removed the matrix determinant example to keep the focus on the general floating-point arithmetic issue.
- Replaced with an example that illustrates floating-point precision in basic arithmetic.
- Added explanation to make the topic more accessible for new users.
[skip azp][skip actions][skip cirrus]

amitsubhashchejara commented:
Yeah, that's what I thought.
So I have updated the example to reflect a more general case of floating-point arithmetic as per your feedback. I removed the determinant-related example and replaced it with the simpler arithmetic case (0.3 - 0.2 - 0.1). Please let me know if further adjustments are needed!

jakevdp reviewed (approved):
Looks good to me!

mattip commented:
Thanks @amitsubhashchejara and reviewers. We can expand on this as time goes by and the need arises.

amitsubhashchejara commented:
Thank you everyone for your assistance throughout.

rkern commented on doc/source/user/basics.types.rst:
```diff
@@ -342,6 +342,30 @@ range of possible values.
+To handle such cases, it's advisable to use functions like `np.isclose` to compare 
+values, rather than checking for exact equality:
+
+    >>> np.isclose(0.3 - 0.2 - 0.1, 0, rtol=1e-05)  # Check for closeness to 0
```
Nitpick for the next time someone looks at this: `rtol` won't really do anything here because one of the comparands is 0. `atol` would be the value that you would want to use to specify your tolerance.

amitsubhashchejara replied:
Actually, it is the default value used by NumPy.

rkern replied:
From the docs:

"""
For finite values, isclose uses the following equation to test whether
two floating point values are equivalent.

 absolute(`a` - `b`) <= (`atol` + `rtol` * absolute(`b`))
"""

Because `b==0`, setting the value of `rtol` won't change anything. In order to demonstrate controlling the tolerance in this case, you must use `atol` in this example and talk about absolute tolerances, not relative tolerances.

amitsubhashchejara replied:
Ok, I will change it.
